src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1766 @@
+        """
+            length normalization
+        """
+        variance = torch.mean(x ** 2, -1, keepdim=True)
+        norm_x = x / torch.sqrt(variance + epsilon)
+        return norm_x
+
+    def _gather_by_expansion(self, vectors, idxs, num_hashes):
+        """
+            expand dims of idxs and vectors for all hashes and gather
+        """
+        expanded_idxs = idxs.unsqueeze(-1).expand(-1, -1, -1, self.attention_head_size)
+        vectors = vectors.repeat(1, 1, num_hashes, 1)
+        return torch.gather(vectors, 2, expanded_idxs)
+
+
+class ReverseSort(Function):
+    """
+        After chunked attention is applied which sorted clusters,
+        original ordering has to be restored.
+        Since customized backward function is used for Reformer,
+        the gradients of the output vectors have to be explicitely
+        sorted here.
+    """
+
+    @staticmethod
+    def forward(ctx, out_vectors, logits, sorted_bucket_idx, undo_sorted_bucket_idx, num_hashes):
+        # save sorted_bucket_idx for backprop
+        ctx.sorted_bucket_idx = sorted_bucket_idx
+        ctx.num_hashes = num_hashes
+
+        out_vectors = out_vectors.detach()
+        logits = logits.detach()
+        # undo sort to have correct order for next layer
+        expanded_undo_sort_indices = undo_sorted_bucket_idx.unsqueeze(-1).expand(out_vectors.shape)
+        out_vectors = torch.gather(out_vectors, 2, expanded_undo_sort_indices)
+        logits = torch.gather(logits, 2, undo_sorted_bucket_idx)
+        return out_vectors, logits
+
+    @staticmethod
+    def backward(ctx, grad_out_vectors, grad_logits):
+        # get parameters saved in ctx
+        sorted_bucket_idx = ctx.sorted_bucket_idx
+        num_hashes = ctx.num_hashes
+
+        # get real gradient shape
+        # shape is BatchSize x NumAttnHeads x ChunkLen * NumHashes
+        grad_logits_shape = grad_logits.shape
+        # shape is BatchSize x NumAttnHeads x ChunkLen * NumHashes x ChunkLen
+        grad_out_vectors_shape = grad_out_vectors.shape
+
+        # split gradient vectors and sorted bucket idxs by concatenated chunk dimension to gather correct indices
+        # shape is BatchSize x NumAttnHeads x NumHashes x ChunkLen
+        grad_logits = torch.reshape(grad_logits, (grad_logits_shape[:2] + (num_hashes, -1)))
+        # shape is BatchSize x NumAttnHeads x NumHashes x ChunkLen x ChunkLen
+        grad_out_vectors = torch.reshape(
+            grad_out_vectors, (grad_out_vectors_shape[:2] + (num_hashes, -1) + grad_out_vectors_shape[-1:])
+        )
+
+        # reshape and expand
+        sorted_bucket_idx = torch.reshape(sorted_bucket_idx, (sorted_bucket_idx.shape[:2] + (num_hashes, -1)))
+        expanded_sort_indices = sorted_bucket_idx.unsqueeze(-1).expand(grad_out_vectors.shape)
+        # reverse sort of forward
+        grad_out_vectors = torch.gather(grad_out_vectors, 3, expanded_sort_indices)
+        grad_logits = torch.gather(grad_logits, 3, sorted_bucket_idx)
```

Comment:
Did you try replacing the `reshape` with `view` here ? The vectors will be used as an argument to `gather` which might not need any contiguous input. 

This would remove some allocations in the bw.